src/main/java/org/opentosca/csarrepo/service/UploadCsarFileService.java

```diff
@@ -85,6 +101,33 @@ private void storeFile(long csarId, InputStream inputStream, String name) {
 
 			String entryDefinition = Extractor.match(Extractor.unzip(temporaryFile, TOSCA_METADATA_FILEPATH),
 					ENTRY_DEFINITION_PATTERN);
+			String xmlData = Extractor.unzip(temporaryFile, entryDefinition);
+
+			DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+			DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+			Document document = documentBuilder.parse(new ByteArrayInputStream(xmlData.getBytes()));
+			XPath xpath = XPathFactory.newInstance().newXPath();
+			XPathExpression expression = xpath.compile(X_PATH_EXPRESSION);
+			NodeList nodeList = (NodeList) expression.evaluate(document, XPathConstants.NODESET);
+
+			Map<String, String> nodeMap = new HashMap<>();
+			for (int i = 0; i < nodeList.getLength(); i++) {
+				nodeMap.put(nodeList.item(i).getNodeName(), nodeList.item(i).getNodeValue());
+			}
+			String serviceTemplateId = nodeMap.get("id");
+			String namespace = nodeMap.get("targetNamespace");
+
+			if (null == csar.getServiceTemplateId()) {
+				csar.setServiceTemplateId(serviceTemplateId);
+				csar.setNamespace(namespace);
+				LOGGER.info("csar: service template id ({}) and namespace ({}) set", serviceTemplateId, namespace);
+			} else if (!csar.getServiceTemplateId().equals(serviceTemplateId)
+					|| (null == csar.getNamespace() && null != namespace && !namespace.equals(null))
```

Comment:
I don't want to get fussy here but I think the !namespace.equals(null) is obsolete here